### PR TITLE
New release v4.0.x

### PR DESCRIPTION
**Pull request by ajgpitch**
Added cython as runtime requirement. Hence mingwpy required also.
patching out cython in windows no longer required
distutils.cfg file added

cython is used at runtime to dynamically build pyx files that are used to create custom functions for ODE RHS

### CODE REVIEW

**Comment by sahmed95** [comment]
Hi @jakirkham 

I am new to conda-forge and I recently joined as a contributor to QuTiP and a maintainer of the recipe. Travis is not able to download Miniconda for Mac in this. We are trying to release qutip 4.0 and this is the new recipe. Is this an issue with conda or are we doing something wrong while writing the recipe ? Please have a look as I am able to download Miniconda with the same commands. Thanks.

**Comment by sahmed95** [comment]
Seems like re-rendering the recipe was required. Please have a look at https://github.com/conda-forge/qutip-feedstock/pull/13

If that works out, this can be closed.

**Comment by ajgpitch** [comment]
@pelson Looks like the tests are having some trouble running. 
@isuruf suggested we rerender, but this did not solve the problems.
As I remember (it was a while ago) the rendering was automatic.

I finally met with the rest of the admin team of the qutip project. We have agreed to make regular releases now, which looks like it might be my job. So, hoping we can get this process nice and smooth :)

**Comment by ajgpitch** [comment]
Ok, so I reinstalled anaconda, installed conda-smithy, created all the tokens, then re-rendered.
Half the macOS packages worked and half failed. The one I looked into had this error during the build.
```
  File "setup.py", line 50, in <module>

    from numpy.distutils.core import setup

  File "/Users/travis/miniconda3/envs/_build_placehold_placehold_placehold_placehold_pla/lib/python3.5/site-packages/numpy/distutils/core.py", line 8, in <module>

    from setuptools import setup as old_setup

ImportError: cannot import name 'setup'
```

Still waiting on appveyor results.

**Comment by sahmed95** [comment]
Do we need setuptools ? All the errors seem to be somewhat related to that. I checked if adding the Lib/disutils/distutils.cfg triggers some need for setuptools but it seems it is okay.

The first warning ["install_requires"](https://travis-ci.org/conda-forge/qutip-feedstock/jobs/186902733#L391) is due to missing setup tools I guess. Similarly the next warning too is due to the same. [Warning 2](https://travis-ci.org/conda-forge/qutip-feedstock/jobs/186902735#L394
)

Something relevant : 

http://stackoverflow.com/questions/8295644/pypi-userwarning-unknown-distribution-option-install-requires

http://stackoverflow.com/questions/32987304/problems-when-converting-from-imp-to-importlib-in-python-3-4

I understand that setuptools is not a default python lib but seems like it is prefered. Check out [scipy's recipe](https://github.com/conda-forge/scipy-feedstock/tree/master/recipe)

This should really have been simpler. :-(

**Comment by ajgpitch** [comment]
When I set this up originally I included setuptools (because I copied the
recipe from elsewhere). While it was still a staged recipe I was advised to
take it out, as we don't use setuptools, but rather we use the numpy
distutils, or something like that.
We do get the occasional warning here and there that the setuptools we are
using are old. Possibly it would make sense to use some newer tools, but
really I don't know for sure. Might not be all that tricky, but probably we
could do with some advice from someone who has some more experience with
these things.

Right now I am working on our pip install, which is also proving awkward.
This distribution business makes quantum mechanics look easy.

Thanks for researching. I will check out your links in a bit.

**Comment by sahmed95** [comment]
Hopefully I can pick up the process soon and then you don't have to deal with this too much. Atleast some of the travis builds are working. Fingers crossed for Appveyor !

Good luck !

**Comment by ajgpitch** [comment]
Looks like the same three packages in the macOS build are failing again. It's a bit annoying, as the the numpy 1.11, py34 & 3.5 are failing, which are probably the ones most people would use.
```
Traceback (most recent call last):

  File "setup.py", line 50, in <module>

    from numpy.distutils.core import setup

  File "/Users/travis/miniconda3/envs/_build_placehold_placehold_placehold_placehold_pla/lib/python3.5/site-packages/numpy/distutils/core.py", line 8, in <module>

    from setuptools import setup as old_setup

ImportError: cannot import name 'setup'
```
It seems like numpy.distutils.core is expecting setuptools. I guess there is no harm in adding setuptools and seeing what happens. AppVeyor is annoyingly slow, so don't know how the Windows builds will fare. However, since the rerender by @isuruf (thank you) it looks like the right packages are being built.

**Comment by ajgpitch** [comment]
So they all built with setuptools included in the build, but all failed on the install, looks like because setuptools needed in run as well if it is used in build

**Comment by sahmed95** [comment]
I initially thought it would be better if we found a workaround to setup
tools but since this is with conda I think it will not be a big issue if we
have setuptools. Conda would definitely have setuptools installed on the
users machine.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -7,27 +7,29 @@ package:
...
 build:
     number: 0
     script: $PYTHON setup.py install  # [unix]
-    skip: True                        # [win and py3k]
+    skip: True                        # [win and py35]
 
 requirements:
   build:
     - python
+    - setuptools
```
Comment (isuruf) [inline recipe/meta.yaml]: Add SciPy here as well

**Comment by isuruf** [comment]
Can you enable CircleCI in your fork, so that Linux tests are run?

**Comment by ajgpitch** [comment]
@isuruf  Sorry I did not see your comment before I made last commit. I have added scipy to the build requirements now.

**Comment by ajgpitch** [comment]
I am not sure how, why or when circleci dropped off. I will have a look at my fork now, see if I can work out what you mean, before I make the next commit

**Comment by isuruf** [comment]
@ajgpitch, this is a CircleCI issue. Earlier you didn't have to do anything. Now the PR author has to enable CircleCI in their fork to make it work. This bug has been reported to CircleCI, but until then you need to enable it manually for your fork.

**Comment by ajgpitch** [comment]
@isuruf Yes, we read something about the issue, but we did not know about this workaround. So thanks for that.
Anyway, looks like it's running now.
Fingers crossed again.

**Comment by ajgpitch** [comment]
@isuruf Looks like everything is building lovely now (at least in linux and macOS, waiting on Windows)
Nothing is being uploaded to the conda-forge yet though. Reports:
```
No BINSTAR_TOKEN present, so no upload is taking place
```
Does the BINSTAR_TOKEN get added when I merge?

Reason I ask is that it is being built against the BLAS version of numpy. Normally we prefer the mkl version. It's not really a requirement, but until I test it I don't know if it will remove the mkl numpy and install the BLAS one when qutip is installed. This would be sub-optimal. I could just add mkl to the build and run requirements, but I don't know if this is recommended.

**Comment by isuruf** [comment]
Yes, the token will be added once you merge. conda-forge uses OpenBLAS, so numpy is built with that. If conda-forge channel has the highest priority, then conda-forge's numpy will be used, removing the numpy from defaults which is built with mkl. If defaults channel is the highest, numpy with mkl is used. Cc @conda-forge/numpy

**Comment by ajgpitch** [comment]
OK, cheers. That sounds fine. We will just wait for AppVeyor now. Hopefully merge tomorrow.

**Comment by ajgpitch** [comment]
All windows packages built.
All numpy 1.11 packages worked. The numpy 1.10 packages failed test with:
```
import: 'qutip'
Traceback (most recent call last):
  File "C:\Miniconda3-x64\conda-bld\test-tmp_dir\run_test.py", line 26, in <module>
    import qutip
  File "C:\Miniconda3-x64\envs\_test\lib\site-packages\qutip\__init__.py", line 205, in <module>
    from qutip.qobj import *
  File "C:\Miniconda3-x64\envs\_test\lib\site-packages\qutip\qobj.py", line 60, in <module>
    from qutip.fastsparse import fast_csr_matrix
  File "C:\Miniconda3-x64\envs\_test\lib\site-packages\qutip\fastsparse.py", line 37, in <module>
    from scipy.sparse.sputils import (upcast, upcast_char, to_native, isdense, isshape,
ImportError: cannot import name 'get_sum_dtype'
```
@nonhermitian might be able to help with a fix for that (only occurs in Windows numpy 1.10)
I would be happy with just the numpy1.11 packages now, and we can apply a fix in the next patch release.
@isuruf Can we merge with these errors?

**Comment by nonhermitian** [comment]
Strange as that func is in scipy.sparse not numpy.  We may not actually
need that func.  Let me check.  At worst case, I can copy the function into
the fastsparse module.

**Comment by ajgpitch** [comment]
Actually I am going to add the fix for this, and fix some other niggles, and then make a 4.0.2 release.

**Comment by isuruf** [comment]
This is because the missing function (`get_sum_dtype`) is in SciPy 0.18 and there's no package for SciPy 0.18 built against numpy 1.10 yet.

**Comment by ajgpitch** [comment]
Ok, so new version 4.0.2 for fixing the `get_sum_dtype` issue. 
Quite a few other fixes in this too.
Tested on Linux and Windows

**Comment by ajgpitch** [comment]
All checks passed! Merging now.

Thanks for help everyone!

**Pull request merged**